### PR TITLE
fix #111346: initial implementation of playback of graceNotesAfter

### PR DESCRIPTION
**Pull request by jimka2001**
implementation to play graceNotesAfter modeled after playing graceNotesBefore.
See also issue https://musescore.org/en/node/111346
for discussion

### CODE REVIEW

**Comment by jimka2001** [comment]
For sure I'll have to also update some midi-out test cases. I'll start looking into that unless there are any serious objections to my general approach.

**Comment by jimka2001** [comment]
hmm. i'm surprised.  All tests passed, that means there's no midi-out test case which contains grace-notes-after.

**Comment by IsaacWeiss** [comment]
Hod does this work for a trilled whole note followed by a grace note after?

**Comment by jimka2001** [comment]
Well it won't work automatically for a trill the trill will play as usual, and the grace note will play as usual.   I have not inserted any extra logic to incorporate the grace-notes-after as a trill termination. 
I do need to verify that the trill playback does not overlap the grace note playback in time.  I haven't verified that yet.

I wish there were an easy way to see the result of midi-out midi-in in a separate tab.   :-( it would make things like this easier to verify.

**Comment by lasconic** [comment]
*I wish there were an easy way to see the result of midi-out midi-in in a separate tab. :-( it would make things like this easier to verify.* 
What do you mean *exactly* ? You can "see" the midi output for a score in pianoroll view. Right click a staff.

**Comment by jimka2001** [comment]
Can the piano roll viewer show musical notation?   If so, it might be
interesting.  But what I'd like know
is what music was created by my code.  I understand that's not something
most people want to see.
Perhaps I can write a function which will midi-out to a temp-file using a
temp-name, then midi-in to another
tab.   I don't know what the export functions look like as far as API, but
it doesn't sound too difficult.

**Comment by jimka2001** [comment]
This commit implements the graceNotesAfter AND graceNotesBefore merged into the articulation such as a trill.  There is still a discussion about whether graceNotesBefore should be merged into the articulation or not, so I don't consider this issue closed.  I can still factor that part out of the algorithm depending on the discussion.

File: libmscore/rendermidi.cpp [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 #include <set>
 
+#include "assert.h"
```
Comment (Jojo-Schmitz) [inline libmscore/rendermidi.cpp]: Really needed? For the Q_ASSERT()?  And not `<assert.h>`

File: libmscore/rendermidi.cpp [hunk 2]
```diff
@@ -360,29 +380,22 @@ static void collectMeasureEvents(EventMap* events, Measure* m, Staff* staff, int
                   Instrument* instr = chord->part()->instrument(tick);
                   int channel = instr->channel(chord->upNote()->subchannel())->channel;
 
-                  foreach (Articulation* a, chord->articulations()) {
+                  foreach (Articulation* a, chord->articulations())
```
Comment (Jojo-Schmitz) [inline libmscore/rendermidi.cpp]: convert to C++11 `for (Articulation* a : chord->articulations())`

File: libmscore/rendermidi.cpp [hunk 3]
```diff
@@ -360,29 +380,22 @@ static void collectMeasureEvents(EventMap* events, Measure* m, Staff* staff, int
...
-                  for (Chord* c : chord->graceNotesBefore()) {
-                        for (const Note* note : c->notes())
+                
+                  if ( !graceNotesMerged(chord))
+                      for (Chord* c : chord->graceNotesBefore())
+                          for (const Note* note : c->notes())
                               collectNote(events, channel, note, velocity, tickOffset);
-                        }
 
                   foreach (const Note* note, chord->notes())
```
Comment (Jojo-Schmitz) [inline libmscore/rendermidi.cpp]: ditto

File: libmscore/rendermidi.cpp [hunk 4]
```diff
@@ -1080,19 +1111,46 @@ bool renderNoteArticulation(NoteEventList* events, Note * note, bool chromatic,
...
+      //    Given a chord from a grace note, (normally the chord contains a single note) and create
+      //    a NoteEvent as if the grace note were part of the articulation (such as trill).  This
+      //    local function works for the graceNotesBefore() and also graceNotesAfter().
+      //    If the grace note has play=false, then it will sound as a rest, but the other grace
+      //    notes will still play.  This means graceExtend simply omits the call to append( NoteEvent(...))
+      //    but still updates ontime +=millespernote.
+      //    RETURNS the new value of ontime, so caller must make an assignment to the return value.
+      auto graceExtend = [millespernote,chord,events] (int notePitch, QVector<Chord*> graceNotes, int ontime) {
+               for (Chord* c : graceNotes) {
+                   for(Note* n: c->notes()) {
```
Comment (Jojo-Schmitz) [inline libmscore/rendermidi.cpp]: space missing between for and (

File: libmscore/rendermidi.cpp [hunk 5]
```diff
@@ -1308,28 +1368,44 @@ void renderGlissando(NoteEventList* events, Note *notestart)
...
 
+    
+    
 //---------------------------------------------------------
 // findFirstTrill
 //  search the spanners in the score, finding the first one
 //  which overlaps this chord and is of type Element::Type::TRILL
 //---------------------------------------------------------
-
+
```
Comment (Jojo-Schmitz) [inline libmscore/rendermidi.cpp]: whitespace damage?

File: libmscore/rendermidi.cpp (hunk 1 above)
Comment (jimka2001) [inline libmscore/rendermidi.cpp]: Yes you're right; we no longer need assert.h.  Thanks.  unfortunately the compiler does not warm me if I include an unnecessary header file.

File: libmscore/rendermidi.cpp (hunk 2 above)
Comment (jimka2001) [inline libmscore/rendermidi.cpp]: Should I convert all these "foreach" usages in the file to "for"?

Comment (Jojo-Schmitz) [inline libmscore/rendermidi.cpp]: Well, at least those you touch or see in the passing, i.e. nearby your changes.

**Comment by jimka2001** [comment]
There are too many small commits.  Before integrating this PR, I'll rebase and squash them together.

**Comment by jimka2001** [comment]
IsaacWeiss, you asked:
> Hod does this work for a trilled whole note followed by a grace note after?
The answer is yes.   That's the plan.

**Comment by Jojo-Schmitz** [comment]
'Hod' is most probably meant to be 'How' and you reply with 'Yes' ?!? ;-)

**Comment by IsaacWeiss** [comment]
@jimka2001, that description sounds perfect. And sorry for the typo (thanks @Jojo-Schmitz!).

**Pull request merged**